broadcastclients/broadcastclients.go

```diff
@@ -140,75 +140,87 @@ func (bcs *BroadcastClients) Start(ctx context.Context) {
+			startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			if _, ok := recentFeedItemsNew[msg.SequenceNumber]; ok {
+				return nil
+			}
+			if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
+				return nil
+			}
+			recentFeedItemsNew[msg.SequenceNumber] = time.Now()
+			if err := router.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+				return err
+			}
+			return nil
+		}
+		confSeqHandler := func(cs arbutil.MessageIndex, router *Router) {
+			startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			if cs == lastConfirmed {
+				return
+			}
+			lastConfirmed = cs
+			if router.forwardConfirmationChan != nil {
+				router.forwardConfirmationChan <- cs
+			}
+		}
+
+		// Two select statements to prioritize reading messages from primary feeds' channels
 		for {
 			select {
 			case <-ctx.Done():
 				return
-
 			// Primary feeds
 			case msg := <-bcs.primaryRouter.messageChan:
-				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
 				primaryFeedIsDownTimer.Reset(MAX_FEED_INACTIVE_TIME)
-				if _, ok := recentFeedItemsNew[msg.SequenceNumber]; ok {
-					continue
-				}
-				if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
-					continue
-				}
-				recentFeedItemsNew[msg.SequenceNumber] = time.Now()
-				if err := bcs.primaryRouter.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+				if err := msgHandler(msg, bcs.primaryRouter); err != nil {
 					log.Error("Error routing message from Primary Sequencer Feeds", "err", err)
 				}
 			case cs := <-bcs.primaryRouter.confirmedSequenceNumberChan:
-				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
 				primaryFeedIsDownTimer.Reset(MAX_FEED_INACTIVE_TIME)
-				if cs == lastConfirmed {
-					continue
-				}
-				lastConfirmed = cs
-				if bcs.primaryRouter.forwardConfirmationChan != nil {
-					bcs.primaryRouter.forwardConfirmationChan <- cs
-				}
-
-			// Secondary Feeds
-			case msg := <-bcs.secondaryRouter.messageChan:
-				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
-				if _, ok := recentFeedItemsNew[msg.SequenceNumber]; ok {
-					continue
-				}
-				if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
-					continue
-				}
-				recentFeedItemsNew[msg.SequenceNumber] = time.Now()
-				if err := bcs.secondaryRouter.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
-					log.Error("Error routing message from Secondary Sequencer Feeds", "err", err)
-				}
-			case cs := <-bcs.secondaryRouter.confirmedSequenceNumberChan:
-				startSecondaryFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
-				if cs == lastConfirmed {
-					continue
-				}
-				lastConfirmed = cs
-				if bcs.secondaryRouter.forwardConfirmationChan != nil {
-					bcs.secondaryRouter.forwardConfirmationChan <- cs
-				}
-
+				confSeqHandler(cs, bcs.primaryRouter)
 			// Cycle buckets to get rid of old entries
 			case <-recentFeedItemsCleanup.C:
```

Comment:
Thats a great point on starving of timers. It makes sense to move cache cleanup to a separate select as it might get starved when feed activity is high. Aiding to the main purpose of `recentFeedItemsCleanup` timer which is to contain the growth of cache.